cmd/tesla-control/commands.go

```diff
@@ -247,40 +265,40 @@ var commands = map[string]*Command{
 		help:             "Unlock vehicle",
 		requiresAuth:     true,
 		requiresFleetAPI: false,
-		handler: func(ctx context.Context, acct *account.Account, car *vehicle.Vehicle, args map[string]string) error {
-			return car.Unlock(ctx)
+		handler: func(ctx context.Context, acct *account.Account, vehicle *vehicle.Vehicle, args map[string]string) error {
```

Comment:
The API should be designed for the consumer, not the developer or the tester, and this API is cumbersome for the consumer.

If we want to abstract the creation of the protobufs, I'd suggest we wrap the code on the other direction:

```go
func (v *Vehicle) Unlock(ctx context.Context) error {
     return v.executeAction(ctx, action.Unlock())
}
```

If the motive here is to improve testability, then this could also be accomplished by implementing a test `connector.Connector` that functions similarly to `net/http/httptest` in that it records requests/responses and you can verify they are as expected.

Reply:
I agree with focusing on consumer vs. author for the API design.
What we lose with moving to this `ExecuteAction` model is that we don't have an easy way to know which actions are available to us, especially because `ExecuteAction` takes a loose `interface{}` type which prevents tools from inferring this statically.
Couple of thoughts:
* Moving the actions out of `Vehicle` and into an `ExecuteAction` has the advantage of clearly separating the domains of what we can do with a vehicle (`Connect`, `Disconnect`, `AddKey`, etc) from all the actions. But we'd need `vehicle` (the package) to specify an `Action` interface to make it easy to statically understand what can be passed in. The implementation details of CarServer vs VSSEC should be kept hidden just like they were before.
* If we keep all the actions straight on `Vehicle` (like Seth suggests), we can test either by having a recorder (like Seth suggests), and/or having an action provider with a big interface that would be swappable for tests. The difference here is that it wouldn't clutter the API from a consumer point of view, and would make things testable a bit more directly than with a recorder. (We could also have the private `executeAction` be a swappable func for tests).